File: core/rss.py
import feedparser
import re

class RSSParser:
    async def __get_latest(self):
        try:
            url = "https://api.mangaupdates.com/v1/releases/rss"
            feed = feedparser.parse(url)
        except:
            try:
                url = "https://api.mangaupdates.com/v1/releases/rss"
                feed = feedparser.parse(url)
            except:
                print("Error: Could not get MangaUpdates RSS feed.")
                return None
        return feed

    async def parse_feed(self):
        feed = await RSSParser.__get_latest(self)
        if feed is None:
            return None
        manga_list = []
        for entry in feed.entries:
            title = entry.title
            try:
                chapter = re.search(r"(v.\d{1,} )?c.\d{1,}(\.\d)?(-\d{1,}(\.\d)?)?", title).group()
                title = title[0: len(title) - len(chapter) - 1]
            except:
                chapter = None
            try:
                scan_group = re.search("(?<=\[).+?(?=\])", title).group()
                title = re.sub("(?<=\[).+?(?=\])", "", title)[3:]
            except:
                scan_group = None
            try:
                if (re.match("^http://", entry.link)):
                    link = re.sub("^http://", "https://", entry.link)
                else:
                    link = entry.link
            except:
                link = None
            
            manga_list.append({"title": title, "chapter": chapter, "scan_group": scan_group, "link": link})
        return manga_list

File: utils/rss.go
package utils

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/mmcdole/gofeed"
	"github.com/valyala/fasthttp"
)

type MangaEntry struct {
	Title     string
	Chapter   string
	ScanGroup string
	Link      string
}

func rssFetchFeed(url string) (*gofeed.Feed, error) {
	req := fasthttp.AcquireRequest()
	defer fasthttp.ReleaseRequest(req)
	req.SetRequestURI(url)

	resp := fasthttp.AcquireResponse()
	defer fasthttp.ReleaseResponse(resp)

	if err := client.Do(req, resp); err != nil {
		return nil, err
	}

	body := resp.Body()
	fp := gofeed.NewParser()
	feed, err := fp.ParseString(string(body))
	if err != nil {
		return nil, err
	}
	return feed, nil
}

func rssGetLatest() (*gofeed.Feed, error) {
	url := "https://api.mangaupdates.com/v1/releases/rss"
	feed, err := rssFetchFeed(url)
	if err != nil {
		feed, err = rssFetchFeed(url)
		if err != nil {
			return nil, fmt.Errorf("Could not get MangaUpdates RSS feed")
		}
	}
	return feed, nil
}

func RssParseFeed() ([]MangaEntry, error) {
	feed, err := rssGetLatest()
	if err != nil {
		return nil, err
	}

	mangaList := []MangaEntry{}
	chapterRegex := regexp.MustCompile(`(v.\d{1,} )?c.\d{1,}(\.\d)?(-\d{1,}(\.\d)?)?`)
	scanGroupRegex := regexp.MustCompile(`$begin:math:display$(.*?)$end:math:display$`)

	for _, entry := range feed.Items {
		title := entry.Title
		chapter := ""
		scanGroup := ""
		link := entry.Link

		chapterMatch := chapterRegex.FindString(title)
		if chapterMatch != "" {
			title = strings.TrimSpace(strings.Replace(title, chapterMatch, "", -1))
			chapter = chapterMatch
		}

		scanGroupMatch := scanGroupRegex.FindStringSubmatch(title)
		if len(scanGroupMatch) > 1 {
			title = strings.TrimSpace(strings.Replace(title, scanGroupMatch[0], "", -1))
			scanGroup = scanGroupMatch[1]
		}

		if strings.HasPrefix(link, "http://") {
			link = "https://" + strings.TrimPrefix(link, "http://")
		}

		mangaList = append(mangaList, MangaEntry{
			Title:     title,
			Chapter:   chapter,
			ScanGroup: scanGroup,
			Link:      link,
		})
	}
	return mangaList, nil
}
